interfaces/builtin/thumbnailer.go

```diff
@@ -0,0 +1,203 @@
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/snapcore/snapd/interfaces"
+)
+
+var thumbnailerServicePermanentSlotAppArmor = []byte(`
+# Description: Allow operating as the thumbnailer service. Reserved because this
+# gives privileged access to the system.
+# Usage: reserved
+
+# DBus accesses
+#include <abstractions/dbus-strict>
+dbus (send)
+    bus=session
+    path=/org/freedesktop/DBus
+    interface=org.freedesktop.DBus
+    member={Request,Release}Name
+    peer=(name=org.freedesktop.DBus),
+
+dbus (send)
+    bus=session
+    path=/org/freedesktop/*
+    interface=org.freedesktop.DBus.Properties
+    peer=(label=unconfined),
+
+# Allow binding the service to the requested connection name
+dbus (bind)
+    bus=session
+    name="com.canonical.Thumbnailer",
+
+# Allow traffic to/from our path and interface with any method
+dbus (receive, send)
+    bus=session
+    path=/com/canonical/Thumbnailer{,/**}
+    interface=com.canonical.Thumbnailer*,
+
+# Allow traffic to/from org.freedesktop.DBus for thumbnailer service
+dbus (receive, send)
+    bus=session
+    path=/com/canonical/Thumbnailer{,/**}
+    interface=org.freedesktop.DBus.*,
+`)
+
+var thumbnailerServiceConnectedPlugAppArmor = []byte(`
+# Description: Allow accessing the thumbnailer service.
+
+# DBus accesses
+#include <abstractions/dbus-session-strict>
+
+# Allow all access to Thumbnailer service
+dbus (receive, send)
+    bus=session
+    peer=(label=###SLOT_SECURITY_TAGS###),
+
+dbus (receive)
+     bus=session
+     path=/
+     interface=org.freedesktop.DBus.ObjectManager
+     peer=(label=unconfined),
+
+# Thumbnailer service interface
+dbus (send)
+     bus=session
+     path="/com/canonical/Thumbnailer"
+     interface="org.freedesktop.DBus.Introspectable"
+     member="Introspect"
+     peer=(label=unconfined),
+
+dbus (send)
+     bus=session
+     path="/com/canonical/Thumbnailer"
+     member={GetAlbumArt,GetArtistArt,GetThumbnail,ClientConfig}
+     peer=(label=unconfined),
+`)
+
+var thumbnailerServicePermanentSlotSecComp = []byte(`
+# Description: Allow operating as the thumbnailer service. Reserved because this
+# gives privileged access to the system.
+# Usage: reserved
+
+getsockname
+recvmsg
+sendmsg
+sendto
+`)
+
+var thumbnailerServiceConnectedPlugSecComp = []byte(`
+# Description: Allow accessing the thumbnailer service.
+# Usage: reserved
+
+getsockname
+recvfrom
+recvmsg
+sendmsg
+sendto
+`)
+
+var thumbnailerServicePermanentSlotDBus = []byte(`
+<policy user="root">
+    <allow own="com.canonical.Thumbnailer"/>
+    <allow send_destination="com.canonical.Thumbnailer"/>
+    <allow send_interface="com.canonical.Thumbnailer"/>
+    <allow send_interface="org.freedesktop.DBus.ObjectManager"/>
+    <allow send_interface="org.freedesktop.DBus.Properties"/>
+</policy>
+<policy context="default">
+    <deny send_destination="com.canonical.Thumbnailer"/>
+</policy>
+`)
```

Comment:
Indeed it sounds like I did it all wrong. Despair.
@jhenstridge are we supposed to have a single instance of the service per user session or is it ok to bundle the thumbnailer service as part of the app and spawn one per app?